src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php

```diff
@@ -1570,6 +1577,54 @@ public function displayCustomizationImageAction(int $orderId, int $type, string
         ]);
     }
 
+    /**
+     * Set order internal note.
+     *
+     * @AdminSecurity(
+     *     "is_granted(['update', 'create'], request.get('_legacy_controller'))",
+     *     redirectRoute="admin_orders_index"
+     * )
+     *
+     * @param $orderId
+     * @param Request $request
+     *
+     * @return Response
+     */
+    public function setInternalNoteAction($orderId, Request $request)
+    {
+        $internalNoteForm = $this->createForm(InternalNoteType::class);
+        $internalNoteForm->handleRequest($request);
+
+        if ($internalNoteForm->isSubmitted()) {
+            $data = $internalNoteForm->getData();
+
+            try {
+                $this->getCommandBus()->handle(new SetInternalOrderNoteCommand(
+                    (int) $orderId,
+                    $data['note']
+                ));
+
+                if ($request->isXmlHttpRequest()) {
+                    return $this->json([
+                        'success' => true,
+                        'message' => $this->trans('Update successful', 'Admin.Notifications.Success'),
+                    ]);
+                }
+
+                $this->addFlash('success', $this->trans('Update successful', 'Admin.Notifications.Success'));
+            } catch (OrderException $e) {
+                $this->addFlash(
+                    'error',
+                    $this->getErrorMessageForException($e, $this->getErrorMessages($e))
```

Comment:
The only missing part now, in the controller, is to update `getErrorMessages` so that it can handle a proper message for the new Constraint exception that you introduced. You would need to add something like this:

```php
            OrderConstraintException::class => [
                OrderConstraintException::INVALID_CUSTOMER_MESSAGE => $this->trans(
                    'The order message given is invalid',
                    'Admin.Orderscustomers.Notification'
                ),
                OrderConstraintException:: INVALID_INTERNAL_NOTE => $this->trans(
                    'The order note given is invalid.',
                    'Admin.Orderscustomers.Notification'
                ),
            ],
```

Although the wording should be validated, I don't know if we should use something more detailed like `The order note given is invalid, please only use clean HTML content.'`

Reply:
I agree with @jolelievre, it'd be more user friendly to have a wording stating why it is invalid. A translation string already exists with this kind of notification, I suggest we use the same and adapt it to this element. Like _Your note contains some invalid HTML and cannot be saved. Please check your content._ in _Admin.Orderscustomers.Notification_. I can suggest it in the code so that you can commit directly, thanks a lot @ks129. ;-)